File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/SessionJsonImpl.java
package com.godaddy.asherah.appencryption;

import org.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.godaddy.asherah.appencryption.envelope.EnvelopeEncryption;
import com.godaddy.asherah.appencryption.utils.Json;

public class SessionJsonImpl<D> implements Session<JSONObject, D> {
  private static final Logger logger = LoggerFactory.getLogger(SessionJsonImpl.class);

  private final EnvelopeEncryption<D> envelopeEncryption;

  /**
   * Creates a new {@code SessionJsonImpl} instance using the provided {@link EnvelopeEncryption} object. An
   * implementation of {@link Session} that encrypts a payload of type {@link org.json.JSONObject}.
   *
   * @param envelopeEncryption An implementation of {@link EnvelopeEncryption} that uses {@link org.json.JSONObject} as
   *                           the Data Row Record format.
   */
  public SessionJsonImpl(final EnvelopeEncryption<D> envelopeEncryption) {
    this.envelopeEncryption = envelopeEncryption;
  }

  @Override
  public JSONObject decrypt(final D dataRowRecord) {
    byte[] jsonAsUtf8Bytes = envelopeEncryption.decryptDataRowRecord(dataRowRecord);

    return new Json(jsonAsUtf8Bytes).toJsonObject();
  }

  @Override
  public D encrypt(final JSONObject payload) {
    byte[] jsonAsUtf8Bytes = new Json(payload).toUtf8();

    return envelopeEncryption.encryptPayload(jsonAsUtf8Bytes);
  }

  @Override
  public void close() {
    try {
      envelopeEncryption.close();
    }
    catch (Exception e) {
      logger.error("unexpected exception during close", e);
    }
  }
}

File: csharp/AppEncryption/AppEncryption/SessionJsonImpl.cs
using System;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;
using GoDaddy.Asherah.AppEncryption.Envelope;
using GoDaddy.Asherah.AppEncryption.Util;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json.Linq;

[assembly: InternalsVisibleTo("AppEncryption.Tests")]
[assembly: System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1816:Call GC.SuppressFinalize correctly", Justification = "This class does not have a finalizer and does not need to suppress finalization.")]

namespace GoDaddy.Asherah.AppEncryption
{
    /// <inheritdoc />
    public class SessionJsonImpl<TD> : Session<JObject, TD>
    {
        private readonly ILogger _logger;
        private readonly IEnvelopeEncryption<TD> envelopeEncryption;

        /// <summary>
        /// Initializes a new instance of the <see cref="SessionJsonImpl{TD}"/> class using the provided
        /// <see cref="IEnvelopeEncryption{TD}"/> object and logger. An implementation of <see cref="Session{TP,TD}"/> that
        /// encrypts a payload of type JObject.
        /// </summary>
        ///
        /// <param name="envelopeEncryption">An implementation of <see cref="IEnvelopeEncryption{TD}"/> that uses
        /// JObject as the Data Row Record format.</param>
        /// <param name="logger">The logger implementation to use.</param>
        public SessionJsonImpl(IEnvelopeEncryption<TD> envelopeEncryption, ILogger logger)
        {
            this.envelopeEncryption = envelopeEncryption;
            this._logger = logger;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="SessionJsonImpl{TD}"/> class using the provided
        /// <see cref="IEnvelopeEncryption{TD}"/> object. An implementation of <see cref="Session{TP,TD}"/> that
        /// encrypts a payload of type JObject. This constructor is provided for backwards compatibility and does not include logging.
        /// </summary>
        ///
        /// <param name="envelopeEncryption">An implementation of <see cref="IEnvelopeEncryption{TD}"/> that uses
        /// JObject as the Data Row Record format.</param>
        public SessionJsonImpl(IEnvelopeEncryption<TD> envelopeEncryption)
            : this(envelopeEncryption, null)
        {
        }

        /// <inheritdoc/>
        public override JObject Decrypt(TD dataRowRecord)
        {
            byte[] jsonAsUtf8Bytes = envelopeEncryption.DecryptDataRowRecord(dataRowRecord);
            return new Json(jsonAsUtf8Bytes).ToJObject();
        }

        /// <inheritdoc/>
        public override TD Encrypt(JObject payload)
        {
            byte[] jsonAsUtf8Bytes = new Json(payload).ToUtf8();
            return envelopeEncryption.EncryptPayload(jsonAsUtf8Bytes);
        }

        /// <inheritdoc/>
        public override async Task<JObject> DecryptAsync(TD dataRowRecord)
        {
            return await Task.FromResult(Decrypt(dataRowRecord));
        }

        /// <inheritdoc/>
        public override async Task<TD> EncryptAsync(JObject payload)
        {
            return await Task.FromResult(Encrypt(payload));
        }

        /// <inheritdoc/>
        public override void Dispose()
        {
            try
            {
                envelopeEncryption.Dispose();
            }
            catch (Exception e)
            {
                _logger?.LogError(e, "unexpected exception during close");
            }
        }
    }
}
